apps/hexagon_dma/process_nv12.cpp

```diff
@@ -0,0 +1,87 @@
+#include <stdio.h>
+#include <memory.h>
+#include <assert.h>
+#include <stdlib.h>
+#include "halide_benchmark.h"
+#include "pipeline_nv12.h"
+#include "HalideRuntimeHexagonDma.h"
+#include "HalideBuffer.h"
+#include "../../src/runtime/mini_hexagon_dma.h"
+
+int main(int argc, char **argv) {
+    if (argc < 3) {
+        printf("Usage: %s width height\n", argv[0]);
+        return 0;
+    }
+
+    const int width = atoi(argv[1]);
+    const int height = atoi(argv[2]);
+
+    // Fill the input buffer with random data. This is just a plain old memory buffer
+    const int buf_size = width * height * 1.5;
+    uint8_t *memory_to_dma_from = (uint8_t *)malloc(buf_size);
+    for (int i = 0; i < buf_size;  i++) {
+        memory_to_dma_from[i] = ((uint8_t)rand()) >> 1;
+    }
+
+    Halide::Runtime::Buffer<uint8_t> input_validation(memory_to_dma_from, width, height, 2);
+    Halide::Runtime::Buffer<uint8_t> input(nullptr, width, height, 2);
+
+    // TODO: We shouldn't need to allocate a host buffer here, but the
+    // current implementation of cropping + halide_buffer_copy needs
+    // it to work correctly.
+    input.allocate();
+
+    // Give the input the buffer we want to DMA from.
+    input.device_wrap_native(halide_hexagon_dma_device_interface(),
+                             reinterpret_cast<uint64_t>(memory_to_dma_from));
+    input.set_device_dirty();
+
+    // In order to actually do a DMA transfer, we need to allocate a
+    // DMA engine.
+    void *dma_engine = nullptr;
+    halide_hexagon_dma_allocate_engine(nullptr, &dma_engine);
+
+    // We then need to prepare for copying to host. Attempting to copy
+    // to host without doing this is an error.
+    // The Last parameter 0 indicate DMA Read
+    halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine, false, eDmaFmt_NV12);
+
+    Halide::Runtime::Buffer<uint8_t> output(width, height, 2);
```

Comment:
You can allocate a buffer of the correct size in a few ways. Allocate a regular buffer (via malloc, etc.), and then create Halide::Runtime::Buffers that point into that buffer, or, something like this is equivalent to what you have now:

    Buffer<uint8_t> output(width, height * 3 / 2);
    Buffer<uint8_t> output_y = output.cropped(1, 0, height);
    Buffer<uint8_t> output_uv = output.cropped(1, height, height / 2);

You can further make this correct by giving uv two channels:

    output_uv.embed(2, 0);
    output_uv.dim(2).extent = 2;
    output_uv.dim(2).stride = 1;
    output_uv.dim(0).extent /= 2;
    output_uv.dim(0).stride = 2;

I'm not completely sure about the last part, but conceptually it should be correct (you want to split the x dimension into an inner dimension with extent 2 and an outer dimension of extent width / 2).